--- origsrc/gst-plugins-bad-1.6.1/configure.ac	2015-10-30 07:11:05.000000000 -0500
+++ src/gst-plugins-bad-1.6.1/configure.ac	2015-11-24 10:24:23.460394600 -0600
@@ -1828,7 +1828,7 @@ AG_GST_CHECK_FEATURE(CURL, [Curl plugin]
     AC_CHECK_HEADERS([unistd.h sys/socket.h sys/types.h netinet/in.h netinet/tcp.h fcntl.h], [ ], [
       AC_CHECK_HEADERS([unistd.h sys/types.h winsock2.h ws2tcpip.h fcntl.h], [ ], [HAVE_CURL="no"])
     ])
-    AC_CHECK_HEADERS([netinet/ip.h], [], [HAVE_CURL="no"],
+    AC_CHECK_HEADERS([netinet/ip.h], [], [],
     [[#ifdef HAVE_SYS_TYPES_H
      # include <sys/types.h>
      #endif
